src/controller/jobRoleController.ts

```diff
@@ -5,16 +5,103 @@ import { bandService } from "../service/bandService";
+    const searchParams = new URLSearchParams();
+    appendURLParams(searchParams, "nameFilter", name);
+    appendURLParams(searchParams, "bandFilter", band);
+    appendURLParams(searchParams, "capabilityFilter", capability);
+    
+    return searchParams.toString();
+}
+
+const decodeURLFilterParams = (req: Request) => {
+    let filters = {} as any;
+    if (req.query?.nameFilter) {
+        filters.nameFilter = decodeURIComponent(req.query.nameFilter as string);
+    }
+    if (req.query?.bandFilter) {
+        const bandFilter = JSON.parse(decodeURIComponent(req.query.bandFilter as string));
+        let bandFilterArray: number[];
+        if (!Array.isArray(bandFilter)) {
+            bandFilterArray = [parseInt(bandFilter)];                    
+        } else {
+            bandFilterArray = bandFilter.map(band => parseInt(band));        
+        }
+        filters.bandFilter = bandFilterArray;
+    }
+    if (req.query?.capabilityFilter) {
+        const capabilityFilter = JSON.parse(decodeURIComponent(req.query.capabilityFilter as string));
+        let capabilityFilterArray: number[];
+        if (!Array.isArray(capabilityFilter)) {
+            capabilityFilterArray = [parseInt(capabilityFilter)];
+        } else {
+            capabilityFilterArray = capabilityFilter.map(capability => parseInt(capability));        
+        }
+        filters.capabilityFilter = capabilityFilterArray;
+    }
+    
+    return filters;
+}
+
 export namespace JobRoles {
     export async function get(req: Request, res: Response): Promise<void> {
+        res.locals.filtered = false;
+
         try {
-            const jobRoles: JobRole[] = await jobRoleService.getJobRoles();
+            let jobRoles: JobRole[] = await jobRoleService.getJobRoles();
+            const { nameFilter, bandFilter, capabilityFilter } = decodeURLFilterParams(req);
+            if (nameFilter) {
+                const lowerCaseFilter = nameFilter.toLowerCase();
+                jobRoles = jobRoles.filter(jobRole => jobRole.jobRoleName.toLowerCase().includes(lowerCaseFilter));
+                res.locals.filtered = true;
+            }
+            if (bandFilter) {
+                jobRoles = jobRoles.filter(jobRole => bandFilter.includes(jobRole.band.bandID));
+                res.locals.filtered = true;
+            }
+            if (capabilityFilter) {
+                jobRoles = jobRoles.filter(jobRole => capabilityFilter.includes(jobRole.capability.capabilityID));
+                res.locals.filtered = true;
+            }
+            
+            res.locals.currentFilterParams = encodeURLFilterParams(nameFilter, bandFilter, capabilityFilter);
             res.locals.jobRoles = jobRoles;
         } catch (e) {
-            res.locals.errorMessage = e;
+            res.locals.errorMessage = e.message;
         }
         res.render("job-roles");
     }
+    
+    export async function getFilter(req: Request, res: Response): Promise<void> {
```

Comment:
would move this and the post into it's own controller. It's feeling complicated in here, and that's an easy way to split it